### PR TITLE
[DateRangeCalendar] Ensure date dragging triggers regardless of trigger element

### PR DESCRIPTION
**Pull request by michelengelen**
This strengthens the date dragging behavior to be less reliant on the specific DOM.

### CODE REVIEW

**Review by flaviendelangle — COMMENTED** [review]
# Code Review: [DateRangeCalendar] Fix dragging start or end date

**PR**: mui/mui-x#21868 by @michelengelen
**Summary**: Fixes drag-and-drop of start/end dates in a selected date range by making element lookups more resilient when drag events target child elements rather than the button itself.

---

## What it does

The core issue is that drag events can target child elements inside `<button>` elements (e.g., a `<span>` inside the button), meaning `event.target` may not have the expected `data-timestamp` or `data-position` attributes. The fix adds `.closest()` fallback lookups to find the nearest ancestor with the expected dataset attribute.

---

## Issues & Suggestions

### 1. Duplicated pattern — extract a helper

The "find element with dataset attribute" pattern is repeated 4 times with slight variations:
```ts
const element = target.dataset.timestamp
  ? target
  : target.closest<HTMLElement>('[data-timestamp]');
```
This should be extracted into a small helper like:
```ts
const closestWithData = (el: HTMLElement, attr: string) =>
  el.dataset[attr] != null ? el : el.closest<HTMLElement>(`[data-${attr}]`);
```
This would reduce duplication and make the intent clearer across all 4 call sites.

### 2. `as any` cast

```ts
return adapter.date(timestamp as any, timezone);
```
This `as any` is suspicious. The old code passed an ISO string (`new Date(timestamp).toISOString()`), the new code passes a raw number. If the adapter's `date()` method doesn't accept a number, this will fail at runtime while silencing the type checker. The PR should either:
- Keep the ISO string conversion if the adapter expects a string, or
- Fix the type signature if `number` is actually valid input

Either way, `as any` should not be the answer here.

### 3. `resolveDateFromTarget` — `instanceof HTMLElement` check

```ts
if (!target || !(target instanceof HTMLElement)) {
  return null;
}
```
This is good defensive coding. However, `instanceof HTMLElement` can fail across iframes/realms. If the date picker is ever rendered inside an iframe, this check would incorrectly return `null`. The codebase guidelines mention using owner utilities for realm-sensitive checks — worth verifying this isn't a concern here.

### 4. Inconsistent element resolution between `resolveDateFromTarget` and `isSameAsDraggingDate`

In `resolveDateFromTarget`, the function first checks `instanceof HTMLElement` and handles `null` gracefully. In `isSameAsDraggingDate`, there's a raw `event.target as HTMLElement` cast with no null/type guard. These should be consistent — if one needs the guard, both likely do.

### 5. `onDragStart` uses `event.currentTarget` while others use `event.target`

In the `onDragStart` handler, the code uses `event.currentTarget`, but in `isSameAsDraggingDate` it uses `event.target`. This inconsistency could lead to different elements being resolved depending on event bubbling. The choice between `target` and `currentTarget` should be deliberate and consistent, or at least documented with a comment explaining why they differ.

### 6. Missing test coverage

The PR description says "fixes dragging of start or end dates" but includes no tests. A test that simulates a drag event targeting a child element of the date button would validate the fix and prevent regression.

---

## Verdict

The fix addresses a real bug (drag events hitting child elements), but the implementation has some rough edges: duplicated lookup patterns, an `as any` cast that hides a potential type mismatch, and no test coverage. Suggest addressing at least items 1, 2, and 6 before merging.












_____________



The `as any` is a deal breaker here, there is no way to assure that all date libraries accept a number on `adapter.date`.

**Comment by LukasTy** [comment]
# PR Review: [DateRangeCalendar] Ensure date dragging triggers regardless of trigger element

**PR:** mui/mui-x#21868
**Author:** @michelengelen (with follow-up commits by @LukasTy)
**Target:** `master` ← `bugfix/date-range-dragging-fix`
**Merge Readiness Score: 9/10**

---

## Problem

Drag events on `DateRangeCalendar` date cells intermittently fail because `event.target` can resolve to a **child element** inside the `<button>` (e.g., text `<span>`, MUI `TouchRipple`) rather than the button itself. Since child elements lack `data-timestamp` and `data-position` attributes, the old code silently failed to resolve dates, causing drag to not initiate.

## Solution

The fix introduces a `.closest()` fallback pattern: when the event target doesn't have the expected data attribute, traverse upward to find the nearest ancestor that does. This is extracted into a reusable `getClosestElementWithDataAttribute` helper. All `event.target` accesses are replaced with proper type-guarded resolution via new shared `domUtils`.

---

## What Changed

| File | Change |
|------|--------|
| `useDragRange.ts` | Core fix — `.closest()` fallback for all target resolutions, hardened type guards, rewritten `resolveButtonElement` |
| `x-internals/domUtils/getTarget.ts` | New utility — resolves event target accounting for shadow DOM (`composedPath`) with fallback |
| `x-internals/domUtils/isHTMLElement.ts` | New utility — cross-realm safe `instanceof HTMLElement` check |
| `DateRangeCalendar.test.tsx` | Two new tests: child-element drag + non-draggable date guard |

## Flavien's Review Feedback — All Addressed

| # | Concern | Resolution |
|---|---------|------------|
| 1 | Extract helper for duplicated `.closest()` pattern | `getClosestElementWithDataAttribute` helper extracted |
| 2 | `as any` cast when passing timestamp to `adapter.date()` | Kept `new Date(timestamp).toISOString()` — no cast needed |
| 3 | `instanceof HTMLElement` fails cross-realm | `isHTMLElement` uses `ownerWindow` fallback |
| 4 | Inconsistent guard between `resolveDateFromTarget` and `isSameAsDraggingDate` | Both now use `isHTMLElement` + `getClosestElementWithDataAttribute` |
| 5 | Inconsistent `target` vs `currentTarget` usage | Comments added; `currentTarget` used deliberately in `handleDragStart`/`handleTouchMove` |
| 6 | Missing test coverage | Two new tests added |

---

## Analysis

### Strengths

1. **Root cause correctly identified.** The intermittent nature maps exactly to varying event targets depending on click position within the button — sometimes hitting the button directly, sometimes a child element.

2. **`isHTMLElement` is well-implemented.** Cross-realm check via `ownerWindow` follows MUI conventions and handles iframe edge cases.

3. **`resolveButtonElement` rewrite is more robust.** Now searches both upward (`.closest('button')`) and downward (BFS), covering both touch scenarios (child targets) and `elementFromPoint` scenarios (wrapper targets).

4. **Good separation of concerns.** DOM utilities placed in `x-internals/domUtils` are reusable across packages. The `exports: { "./*": ... }` wildcard in `x-internals/package.json` automatically supports the new subpath.

5. **`getTarget` handles edge cases.** The `composedPath()[0] ?? event.target` fallback guards against empty composed paths.

6. **Tests deterministically exercise the fix.** Synthetic `<span>` children are appended to buttons, then drag events are fired on them — this guarantees the `.closest()` path is hit rather than silently falling back to the button itself.

### Remaining Nits

1. **`resolveButtonElement` BFS is unbounded.** The breadth-first downward search has no depth limit. Not a real concern for date cells (small DOM subtrees), but worth noting if this utility is later reused in deeper DOM contexts.

2. **`handleDragOver` behavioral change.** The old code called `setRangeDragDay(resolveDateFromTarget(event.target, ...))` in `handleDragOver`. The new code only sets `dropEffect` there, relying on `handleDragEnter` for state updates. This is correct (`dragEnter` fires first and is the right place for state changes), but it's an incidental behavioral change unrelated to the child-element fix — worth a note in the PR description.

Neither of these blocks merging.

## Verdict

The fix correctly addresses the root cause of intermittent drag failures. All of Flavien's review feedback has been incorporated. New `domUtils` are clean and reusable. Tests deterministically validate the `.closest()` code path. Ready to merge.

File: packages/x-internals/src/domUtils/getTarget.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+/**
+ * Returns the target element of an event, accounting for shadow DOM.
+ * @param event The event object.
+ * @returns The target element of the event.
+ */
+export function getTarget(event: Event) {
```
Comment (LukasTy) [inline packages/x-internals/src/domUtils/getTarget.ts:6]: This is pretty much a straight copy from floating ui dom utils, and the below `isHTMLElement` is inspired by it.

**Review by flaviendelangle — APPROVED** [review]
# PR Review: [DateRangeCalendar] Ensure date dragging triggers regardless of trigger element

**PR:** mui/mui-x#21868
**Author:** michelengelen
**Base:** master <- bugfix/date-range-dragging-fix
**Size:** +181 / -22 across 5 files

---

## Overview

Fixes a bug where dragging the start/end dates of a selected range fails when the drag event targets a child element (e.g., text span, TouchRipple) inside the day button rather than the button itself. The fix uses `.closest()` to walk up the DOM tree and find the element with `data-timestamp`.

---

## What Changed

| File | Change |
|------|--------|
| `useDragRange.ts` | Core fix: use `.closest()` to resolve `data-timestamp` from child elements; new `getClosestElementWithDataAttribute` and `isEnabledButtonElement` helpers; `resolveButtonElement` now searches up (`.closest('button')`) and down (BFS) |
| `getTarget.ts` | New shared utility: resolves event target accounting for shadow DOM via `composedPath()` |
| `isHTMLElement.ts` | New shared utility: cross-realm `instanceof HTMLElement` check |
| `domUtils/index.ts` | Barrel export for the two new utilities |
| `DateRangeCalendar.test.tsx` | Two new tests: child-element drag scenario + non-draggable date guard |

---

## Strengths

- **Root cause correctly identified**: The real issue is `event.target` pointing to a child element (span, ripple) that lacks `data-timestamp`. Using `.closest()` is the idiomatic DOM fix.
- **Shared utilities in `x-internals`**: `getTarget` (shadow DOM aware) and `isHTMLElement` (cross-realm safe) are well-scoped, reusable, and correctly implemented.
- **Good test coverage**: The child-element drag test directly exercises the `.closest()` fallback path. The non-draggable date test adds a useful guard.
- **Defensive coding**: Null checks throughout, optional chaining on `buttonDataset?.timestamp`, proper type narrowing with `isHTMLElement`.

---

## Issues & Suggestions

### Medium

1. **`getClosestElementWithDataAttribute` uses camelCase for `dataAttribute` but constructs a kebab-case selector**
   ```ts
   element.closest<HTMLElement>(`[data-${dataAttribute}]`)
   ```
   This is called with `'timestamp'` and `'position'` (single words), so it works. But the first branch checks `element.dataset[dataAttribute]` (camelCase lookup) while `.closest()` uses `data-${dataAttribute}` (kebab-case selector). For multi-word attributes like `dataFooBar`, these would diverge (`dataset.fooBar` vs `[data-fooBar]` which should be `[data-foo-bar]`). Currently safe since all callers use single words, but the asymmetry is a latent bug. Consider documenting that `dataAttribute` must be a single lowercase word, or normalizing both paths.

2. **`resolveButtonElement` BFS could be expensive on deep DOM trees**
   The old recursive approach only walked first-children (depth-first, leftmost path). The new BFS traverses *all* descendants. For typical day cells this is fine (shallow DOM), but it's a behavioral change worth noting. Consider adding a depth limit or documenting the expected DOM structure.

3. **`isHTMLElement` cross-realm check could throw**
   ```ts
   value instanceof (ownerWindow(value as Node) as ...).HTMLElement
   ```
   If `ownerWindow` returns a detached/destroyed window (e.g., removed iframe), accessing `.HTMLElement` may throw. A try/catch would make this bulletproof.

### Low

4. **Test uses `any` for date tuple type**
   ```ts
   const initialValue: [any, any] = [...]
   ```
   Could use the actual date type from the adapter for better type safety, though this follows existing test conventions.

5. **`getTarget` fallback comment mentions "browsers without shadow DOM"** - This is increasingly rare (all modern browsers support it). The fallback is still correct but the comment could be simplified.

6. **No barrel export from `x-internals/src/index.ts`** - The new `domUtils` module is imported directly (`@mui/x-internals/domUtils`). Verify this path is configured in the package's `exports` map in `package.json`, otherwise it won't resolve for consumers.

---

## Rating: 4 / 5

Solid bug fix with good test coverage and well-structured shared utilities. The `.closest()` approach is correct and idiomatic. Loses a point for the camelCase/kebab-case asymmetry in `getClosestElementWithDataAttribute` (latent bug for multi-word attributes) and the missing safety net in the cross-realm `isHTMLElement` check. Overall, this is a clean, well-motivated PR that properly addresses a real user-facing bug.

File: packages/x-internals/src/domUtils/isHTMLElement.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  if (value instanceof HTMLElement) {
+    return true;
+  }
+  // Cross-realm: must be an element node (nodeType 1) from another window.
+  // The try/catch guards against detached/destroyed windows where accessing
+  // `.HTMLElement` on the owner window may throw.
+  try {
+    return (
+      (value as Node).nodeType === 1 &&
+      value instanceof (ownerWindow(value as Node) as Window & typeof globalThis).HTMLElement
```
Comment (romgrk) [inline packages/x-internals/src/domUtils/isHTMLElement.ts:19]: Doesn't the `instanceof ownerWindow(...).HTMLElement` imply that `nodeType == 1`? Why do we need to check `nodeType` only for cross-realm checks?

Comment (michelengelen) [reply packages/x-internals/src/domUtils/isHTMLElement.ts:19]: You're right that `instanceof ownerWindow(...).HTMLElement` would be sufficient. The `nodeType === 1` check is a short-circuit optimization to avoid calling `ownerWindow()` on obvious non-element nodes, since `ownerWindow` traverses the DOM to find the owner window, which is more expensive than a simple property read.

**Review by LukasTy — APPROVED** [review]
